### PR TITLE
Enable Multi-arch building with a Unified Dockerfile

### PR DESCRIPTION
**Pull request by TrevorTaoARM**
Now we try to enable a unified Dockerfile to build images
for multiple architectures(here amd64 and arm64), it's now
for the apiserver component here.
With more agreements, the way used in this PR can be
popularized to other Calico components.

To use a single Dockerfile for image building would greatly benefit the
maintenance and development work for us. Here we think for most
cases a unified Dockerfile can work with a unified view.

Signed-off-by: TrevorTaoARM <trevor.tao@arm.com>

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

**Review by caseydavenport — CHANGES_REQUESTED** [review]
One comment on the details.

But, at a high level, I wonder if this is worth doing. For many of our other directories, there are substantial differences between the images for each ARCH, and I worry that this change will not be something we can roll out to every image, leaving us in a place where we have inconsistent methods for multi-arch for different images.

**Comment by frozenprocess** [comment]
> I worry that this change will not be something we can roll out to every image.

I think it will cause chaos when we get to `calico-node` dockerfile.
From the top of my mind I would say different package names 
[arm64](https://github.com/projectcalico/calico/blob/78a071caf0c5d5c3212e861974da3055e1415844/node/Dockerfile.arm64#L155) [amd64](https://github.com/projectcalico/calico/blob/78a071caf0c5d5c3212e861974da3055e1415844/node/Dockerfile.amd64#L145)
different cleanup steps 
[arm64](https://github.com/projectcalico/calico/blob/78a071caf0c5d5c3212e861974da3055e1415844/node/Dockerfile.arm64#L212)

[amd64](https://github.com/projectcalico/calico/blob/78a071caf0c5d5c3212e861974da3055e1415844/node/Dockerfile.amd64#L211)

and more :D

File: apiserver/Makefile [hunk 1]
```diff
@@ -58,6 +58,9 @@ RELEASE_LDFLAGS = -ldflags "$(VERSION_FLAGS) -w"
 
 include ../lib.Makefile
 
+# Now we support 2 architectures: amd64, arm64
+ENABLEDARCHES ?= amd64 arm64
```
[the remark this answers is not in the document]
Comment (TrevorTaoARM) [reply apiserver/Makefile:62]: Yes, I knew VALIDARCHES are used by many other components and build logics, which just extracts architecture tag from the extensions of Dockerfile.xxx, so I added a new variable named ENABLEDARCHES to distinguish between them, and it's set and takes effect just in the apiserver here.
For other components(e.g., node,...), the tag ENABLEDARCHES can also be used to indicate the unified cases, if needed in the future.

**Comment by TrevorTaoARM** [comment]
> > I worry that this change will not be something we can roll out to every image.
> 
> I think it will cause chaos when we get to `calico-node` dockerfile. From the top of my mind I would say different package names [arm64](https://github.com/projectcalico/calico/blob/78a071caf0c5d5c3212e861974da3055e1415844/node/Dockerfile.arm64#L155) [amd64](https://github.com/projectcalico/calico/blob/78a071caf0c5d5c3212e861974da3055e1415844/node/Dockerfile.amd64#L145) different cleanup steps [arm64](https://github.com/projectcalico/calico/blob/78a071caf0c5d5c3212e861974da3055e1415844/node/Dockerfile.arm64#L212)
> 
> [amd64](https://github.com/projectcalico/calico/blob/78a071caf0c5d5c3212e861974da3055e1415844/node/Dockerfile.amd64#L211)
> 
> are some of the basic stuff to worry about.

I recognize there are many differences from the stuffs of Calico between the 2 architectures. But it's still can be solved and under control from the technical side.
Actually, many projects just use unified Dockerfiles which should be as complex as Calico, e.g., Istio, all its components (build-tools, pilot, proxyv2, operator, ...) and you could refer the following samples:
https://github.com/istio/tools/blob/master/docker/build-tools/Dockerfile
https://github.com/istio/istio/blob/master/pilot/docker/Dockerfile.proxyv2

I also checked and built calico-node Dockerfiles, there are differences between arches, but most of them are the same and I didn't think those differences can not be merged in a single Dockerfile with other unified scripts(e.g., cleanup).

**Comment by TrevorTaoARM** [comment]
> 

I checked the Dockerfiles of different arches of various Calico components, and found most contents of them are  the same between arches, and I think those can be differentiated with some tags(TARGETPLATFORM, TARGETARCH, if...) in the Dockerfile/scripts to enable a unified one.
Actually, many projects which should be as complex as Calico just use unified Dockerfiles , e.g., Istio(https://github.com/istio), all its components (build-tools, pilot, proxyv2, operator, ...) just use a unified Dockerfile, and you could refer the following samples:
https://github.com/istio/tools/blob/master/docker/build-tools/Dockerfile
https://github.com/istio/istio/blob/master/pilot/docker/Dockerfile.proxyv2

File: apiserver/Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (caseydavenport) [reply apiserver/Makefile:62]: Right, but what I meant is that VALIDARCHES is actually still needed within this Makefile. 

We import `lib.Makefile` into this Makefile, and several of the helper functions from there expect VALIDARHCES to exist and be set properly.

**Comment by mgleung** [comment]
Hi @TrevorTaoARM , just wanted to bump this again since it seems that this PR has hit a block. From my understanding, we have a couple of issues in order to get this in:
1. We need to overwrite `VALIDARCHES` instead of creating a new variable
2. Unified dockerfiles are great for some of our simpler images (including the apiserver here) but may add a lot of confusion (regarding readability) for more complex images (like node). I think that this difference will be a point of contention just because it's more opinion based, but I noticed that the referenced Istio images only seem to support `amd64` and `arm64` builds. Do you have an idea if this will work for some of the other archs we support (`ppc64le`, `s390x`)? I'm not sure how much those archs may end up differing in the future which may make our Dockerfiles unwieldy.

**Comment by TrevorTaoARM** [comment]
> Hi @TrevorTaoARM , just wanted to bump this again since it seems that this PR has hit a block. From my understanding, we have a couple of issues in order to get this in:
> 
> 1. We need to overwrite `VALIDARCHES` instead of creating a new variable
> 2. Unified dockerfiles are great for some of our simpler images (including the apiserver here) but may add a lot of confusion (regarding readability) for more complex images (like node). I think that this difference will be a point of contention just because it's more opinion based, but I noticed that the referenced Istio images only seem to support `amd64` and `arm64` builds. Do you have an idea if this will work for some of the other archs we support (`ppc64le`, `s390x`)? I'm not sure how much those archs may end up differing in the future which may make our Dockerfiles unwieldy.

1. Yes, we can overwrite "VALIDARCHES" instead of "ENABLEDARCHES" in the future, but now they are still used by other components, so currently I add the "ENABLEDARCHES" here, and if the Dockerfiles of all components have been multi-arched in the future, we can return to "VALIDARCHES" easily.
2. I think unified Dockerfiles are a necessary trend which is much easier to maintain, for the complex images(like node) that you mentioned, I think the Dockerfile of istio/build-tools look like more complex than Node:
https://github.com/istio/tools/blob/master/docker/build-tools/Dockerfile
For the support of othe arches(ppc64le, s390x), I didn't see too much difference from amd64 and arm64, and I think the most multi-arch work for amd64/arm64 can also be used on other arches, e.g., ppc64le, if there are some special treatment needed for those arches, I think we can do it in the same way when we meet for amd64 and arm64.
Since I had no ppc64le/s390x platforms in my local environment, I can't test for those arches, but I think other guys who have those platforms may help here if needed.

**Comment by hjiawei** [comment]
This is addressed as part of https://github.com/projectcalico/calico/pull/8299.

**Pull request closed without merging**